### ISSUE TITLE
R session crashes when trying to read (some) text values from a Stata file with `read_dta()`

### ISSUE THREAD

**Issue by DaniMori** [tidyverse/haven#620]
I have a dataset in a `.dta` file which opens normally in Stata, and is also read normally with `read_dta()` in haven v.2.3.1.
However, when testing in versions 2.4.0 and 2.4.1, an "R Session Aborted" crash is produced.

I can't provide a "reprex", as when I try to reprex it, I obtain the message:

"Error: This reprex appears to crash R"

So here is the code to reproduce it:

```r
library(tidyverse)
library(haven)

setwd("~/../OneDrive - UAM/Workspace/R_bug_reporting_and_help")

# Both one-variable datasets are read normally
q4534 <- read_dta("data/q4534.dta")
q4554 <- read_dta("data/q4554.dta")

# If I collapse them together, write them as a .dta file and read it again,
#   it works normally
both_vars <- bind_cols(q4534, q4554)
write_dta(both_vars, path = "data/test.dta")
read_dta("data/test.dta")

# However, the dataset with the original two-variable dataset crashes:
read_dta("data/q4534_q4554.dta")
```

The three files are attached. Note that there are two variables with two cases each, and both are text values. The bug seems to happen when both are text variables; also, you may notice my working directory is a cloud storage drive (OneDrive). I tried pausing the synchronization, in case that could be causing the problem, but it happened anyway.

[data.zip](https://github.com/tidyverse/haven/files/6878503/data.zip)

**Comment by wendimalley** [comment]
I am having this same issue.  I am also using a cloud storage drive.

**Comment by gorcha** [comment]
Hi,

This should already be fixed in the current development version - try installing from github and see if that fixes the issue!
```r
devtools::install_github("tidyverse/haven")
```

**Comment by DaniMori** [comment]
It seems to work now, as it does not terminate the R session abnormally. However, look at the resulting tibble read:

````
read_dta("data/q4534_q4554.dta")
# A tibble: 2 x 2
  q4534 q4554 
  <chr> <chr> 
1 " "   "text"
2 ""    " "   
````

Element in column `q4535`, row 2, is "", an empty (0-character) string, where it should be "text" (please see dataset "data/q4534_q4554.dta" in `data.zip`, attached to the first message of the issue).

The scary thing is... I tried with version 2.3.1, and apparently this example doesn't work either! It gives a `NA` where the non-empty string should be:

````
read_dta("data/q4534_q4554.dta")
# A tibble: 2 x 2
  q4534 q4554 
  <chr> <chr> 
1 " "   "text"
2  NA   " " 
````
I also explored what happened in other non-dummy, actual datasets, and I have found at least that some variables have all 0-length character values where character strings should be.

FYI @wendimalley , I tried copying my project elsewhere in my hard drive and I can assert for sure that it was not due to cloud storage syncing.

**Comment by gorcha** [comment]
Hi @DaniMori, it looks like this is because of a bug in readstat (the library behind haven), which was fixed very recently but hasn't made its way through to a release yet.

@hadley this is a pretty hectic bug but has a simple fix (see WizardMac/ReadStat#248) so it's probably worth incorporating the change before the readstat release officially has it? Sorry I should have mentioned it in the PR the other day.

**Comment by DaniMori** [comment]
Many thanks for the clarifications. Seems this issue was already reported in #600 , I have only found out now after seeing the issue in readstat, my apologies.
